com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorBase.cs

```diff
@@ -0,0 +1,336 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using UnityEngine;
+using Unity.MLAgents.Sensors;
+using UnityEngine.Profiling;
+using Object = UnityEngine.Object;
+
+[assembly: InternalsVisibleTo("Unity.ML-Agents.Extensions.TestUtils")]
+namespace Unity.MLAgents.Extensions.Sensors
+{
+    /// <summary>
+    /// Grid-based sensor.
+    /// </summary>
+    public class GridSensorBase : ISensor, IBuiltInSensor, IDisposable
+    {
+        string m_Name;
+        Vector3 m_CellScale;
+        Vector3Int m_GridSize;
+        int[] m_ChannelDepths;
+        string[] m_DetectableObjects;
+        SensorCompressionType m_CompressionType;
+        ObservationSpec m_ObservationSpec;
+        internal BoxOverlapChecker m_BoxOverlapChecker;
+
+        // Buffers
+        float[] m_PerceptionBuffer;
+        Color[] m_PerceptionColors;
```

Comment:
Now that you're separating out the sensor for gizmo drawing, can you move this into a subclass of GridSensorBase?

Reply:
I don't quite understand this one - m_PerceptionBuffer and m_PerceptionColors are both used in general grid sensors. The data will go from `m_PerceptionBuffer` (float) -> `m_PerceptionColors (Colors)` -> `m_PerceptionTexture.SetPixels(m_PerceptionColors)`

Reply:
My misunderstand - I thought m_PerceptionColors was only used for gizmos.